### PR TITLE
Clarify concurrency requirements for trace and metrics

### PR DESCRIPTION
**Pull request by bogdandrutu**
Fixes https://github.com/open-telemetry/opentelemetry-specification/issues/662

## Changes

Add concurrency requirements for metrics API and polish a bit trace API requirements.

### CODE REVIEW

File: specification/concurrency.md [hunk 1]
```diff
@@ -18,5 +22,18 @@ SpanBuilder is used by more than one thread/coroutine.
...
+
+## Metrics
+
+**MeterProvider** - all methods are safe to be called concurrently.
+
+**Meter** - all methods are safe to be called concurrently.
+
+**Instrument** - All methods of any Instrument are safe to be called concurrently.
+
+**Bound** - All methods of any Bound are safe to be called concurrently.
```
Comment (jkwatson) [inline specification/concurrency.md:39]:
```suggestion
**Bound Instrument** - All methods of any Bound Instrument are safe to be called concurrently.
```

**Comment by bogdandrutu** [comment]
@jkwatson you like making "requested changes" reviews :)

**Comment by jkwatson** [comment]
> @jkwatson you like making "requested changes" reviews :)

I actually hate it, but the github UI often will override my desires. :(

File: specification/concurrency.md [hunk 2]
```diff
@@ -18,5 +22,18 @@ SpanBuilder is used by more than one thread/coroutine.
 
 **Span** - All methods of Span are safe to be called concurrently.
 
+**Event** - Events are immutable and is safe to be used concurrently. Lazy
+initialized eventss must be implemented to be safe to be called concurrently.
```
Comment (jtescher) [inline specification/concurrency.md:26]:
```suggestion
initialized events must be implemented to be safe to be called concurrently.
```

File: specification/concurrency.md [hunk 3]
```diff
@@ -18,5 +22,18 @@ SpanBuilder is used by more than one thread/coroutine.
 
 **Span** - All methods of Span are safe to be called concurrently.
 
+**Event** - Events are immutable and is safe to be used concurrently. Lazy
```
Comment (yurishkuro) [inline specification/concurrency.md:25]:
```suggestion
**Event** - Events are immutable and safe to be used concurrently. Lazy
```

File: specification/concurrency.md [hunk 4]
```diff
@@ -9,6 +9,10 @@ situations.
 The following are general recommendations of concurrent call safety of
 specific subsets of the API.
 
+## Trace API
+
+**TracerProvider** - all methods are safe to be called concurrently.
```
Comment (Oberon00) [inline specification/concurrency.md:14]: Also shutdown?

File: specification/concurrency.md [hunk 5]
```diff
@@ -18,5 +22,18 @@ SpanBuilder is used by more than one thread/coroutine.
 
 **Span** - All methods of Span are safe to be called concurrently.
 
+**Event** - Events are immutable and safe to be used concurrently. Lazy
+initialized events must be implemented to be safe to be called concurrently.
```
Comment (Oberon00) [inline specification/concurrency.md:26]: Really? Is there a case where the same event would be initialized more than once? Maybe multiple exporters?

Comment (bogdandrutu) [reply specification/concurrency.md:26]: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#add-events

If an interface is defined by the language calls to that API should be thread safe. Also this is similar with existing Link

File: specification/concurrency.md (hunk 4 above)
Comment (bogdandrutu) [reply specification/concurrency.md:14]: It should be thread safe (no memory corruption or other problems related to thread safety), behavior is a different story, probably should be a no-op the second time, but I believe it should be safe and user should not implement a synchronization mechanism to ensure called only once.

**Comment by bogdandrutu** [comment]
@Oberon00 indeed we can do a follow up and move every section to the corresponding API file

File: specification/concurrency.md (hunk 5 above)
Comment (Oberon00) [reply specification/concurrency.md:26]: Then I suggest

```suggestion
initialized events must be thread safe. This is the responsibility of the implementer of these events.
```

Comment (bogdandrutu) [reply specification/concurrency.md]: Done.

**Pull request merged**